packages/neon_framework/packages/notifications_push_repository/lib/src/notifications_push_repository.dart

```diff
@@ -0,0 +1,243 @@
+  }
+
+  Future<void> _unregisterUnifiedPush() async {
+    final subscriptions = await _storage.readSubscriptions();
+    for (final entry in subscriptions.entries) {
+      final accountID = entry.key;
+      final account = _accounts.firstWhereOrNull((account) => account.id == accountID);
+      var subscription = entry.value;
+
+      subscription = await _unregisterNextcloud(accountID, account, subscription);
+      if (subscription.endpoint != null) {
+        _log.finer('Unregistering $accountID from UnifiedPush');
+
+        subscription = subscription.rebuild((b) => b.endpoint = null);
+        await _saveUpdatedSubscription(subscriptions, accountID, subscription);
+
+        await UnifiedPush.unregister(accountID);
+      } else {
+        await _saveUpdatedSubscription(subscriptions, accountID, subscription);
+      }
+    }
+  }
+
+  Future<PushSubscription> _registerNextcloud(Account account, String endpoint, PushSubscription subscription) async {
+    _log.finer('Registering ${account.id} at Nextcloud');
+
+    try {
+      final response = await account.client.notifications.push.registerDevice(
+        $body: notifications.PushRegisterDeviceRequestApplicationJson(
+          (b) => b
+            ..pushTokenHash = notifications.generatePushTokenHash(endpoint)
+            ..devicePublicKey = _privateKey.publicKey.toFormattedPEM()
+            ..proxyServer = '$endpoint#',
+        ),
+      );
+
+      return subscription.rebuild(
+        (b) => b
+          ..endpoint = endpoint
+          ..pushDevice.replace(response.body.ocs.data),
+      );
+    } on DynamiteApiException catch (error) {
+      _log.warning('Failed to register ${account.id} at Nextcloud', error);
+    }
+
+    return subscription;
+  }
+
+  Future<PushSubscription> _unregisterNextcloud(
+    String accountID,
+    Account? account,
+    PushSubscription subscription,
+  ) async {
+    if (subscription.pushDevice != null) {
+      _log.finer('Unregistering $accountID from Nextcloud');
+
+      try {
+        await account?.client.notifications.push.removeDevice();
+      } on DynamiteApiException catch (error) {
+        _log.warning('Failed to unregister $accountID at Nextcloud', error);
+      }
+    }
+
+    // Remove the push device either way, as we need to register for a new one later
+    return subscription.rebuild((b) => b.pushDevice = null);
+  }
+
+  Future<void> _saveUpdatedSubscription(
+    BuiltMap<String, PushSubscription> subscriptions,
+    String accountID,
+    PushSubscription subscription,
+  ) async {
+    await _storage.saveSubscriptions(
+      subscriptions.rebuild(
+        (b) {
+          if (subscription.endpoint == null && subscription.pushDevice == null) {
+            _log.finer('Removing subscription for $accountID');
+            b.remove(accountID);
+          } else {
+            _log.finer('Saving subscription for $accountID');
+            b[accountID] = subscription;
+          }
+        },
+      ),
+    );
+  }
+}
```

Comment:
This would fit better into the storage layer, but we can also keep it here until #1538 has landed.

Reply:
I was thinking about that, not sure why I didn't do it in the end.